### PR TITLE
[#144] Correct user update

### LINKED ISSUE #144: Lost all my workflows when updating my profile
# Objective

Not losing my workflows when updating my profile.

# Current situation

* Create a workflow
* Edit your profile (any field)
* Go to the list of workflows
* Reload
* ==> No workflow remain

# Expected situation

* Create a workflow
* Edit your profile (any field)
* Go to the list of workflows
* Reload
* ==> All workflows remain

# More information

Tested with a non-google user.

### PR DESCRIPTION
**Pull request by grizio**
When updating the user, some data were lost.
The reason was the update is too much permissive and the client gave not all data.
We change a little the design to only accept wanted values.

This commit does the following:

* Introduce `io-ts` (fully js compatible)
* Create a new route with a patch system
* Create the model validation for this new route
* Create the new service to patch the user
* Adapt client-side to use this new route

This PR resolves #144

### CODE REVIEW

**Comment by simonLouvet** [comment]
hello @grizio . before merge, i don't see where so you check the meil regex.

**Comment by grizio** [comment]
I added the label `DO NOT MERGE YET` because there are some things that need to be done. The mail is currently read-only but I should ensure we do not update it on the back. It will take some time and analysis before finishing this PR (normally by the end of this week).

File: core/lib/user_lib.js [hunk 1]
```diff
@@ -309,6 +310,31 @@ function _userGraph(userId) {
   });
 } // <= _userGraph
 
+/**
+ * @param {string} id
+ * @param {ProfilPatch} userPatch
+ * @return {Promise<UserDocument>}
+ */
+function updateProfil(id, userPatch) {
```
Comment (grizio) [inline core/lib/user_lib.js]: At this time, the model is already validated. After investigation on previous code, we do not need to check other things.

File: main/server/userWebservices.js [hunk 2]
```diff
@@ -97,4 +99,13 @@ module.exports = function (router) {
       next(e)
     })
   })
+
+  router.patch('/users/me',
```
Comment (grizio) [inline main/server/userWebservices.js]: We create another route so we do not break eventually existing code depending on `put` version (I did not see one though).

File: main/server/validations/index.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Validate and normalize the input (body) of the request.
+ * It ensures the body has the right model and does not include unnecessary attributes (they will be stripped)
+ *
+ * @template {A}
+ * @param {Decoder<A>} type
+ * @return {Function}
+ */
+function validateRequestInput(type) {
```
Comment (grizio) [inline main/server/validations/index.js]: This helper function helps us to validate input model so we are sure it respects what we expect in the request processing.

File: main/server/validations/userValidations.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * @typedef {{
+ *   name: string,
+ *   society?: string,
+ *   job?: string
+ * }} ProfilPatch
+ */
+
+/** @type {Decoder<ProfilPatch>} */
+const userPatchType = t.exact(t.type({
```
Comment (grizio) [inline main/server/validations/userValidations.js]: We ensure to have an exact model and avoid invalid data / data injection trough the rest of the process.

File: main/server/validations/common.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
+const t = require('io-ts')
```
Comment (grizio) [inline main/server/validations/common.js]: Lots of helpers to ensure valid data, can be seen as complex (and it is :disappointed: ) but then we can only use them.

File: main/client/static/tag/profil.tag [hunk 6]
```diff
@@ -154,23 +154,7 @@
...
-          RiotControl.trigger('update_user', {
-            user: this.profil,
-            mailChange: false
-          });
-        }
-      } else {
-        this.result = false;
-        this.resultEmail = "L'email n'est pas au bond format";
-      }
+      RiotControl.trigger('update_user', this.profil);
```
Comment (grizio) [inline main/client/static/tag/profil.tag]: The email field is read only and not used on server side so we can remove this useless code.

**Comment by simonLouvet** [comment]
@grizio a merged but regex still avaible in tag login.tag. can you remove it?

**Comment by grizio** [comment]
I took some time to check. The email in `login.tag` seems still valid. Here, I removed the regex here because the email is never updated.

**Pull request merged**